lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -991,16 +1029,69 @@ instance FromJSON ApiEpochInfo where
 instance ToJSON ApiEpochInfo where
     toJSON = genericToJSON defaultRecordTypeOptions
 
-instance DecodeAddress n => FromJSON (ApiSelectCoinsData n) where
+instance FromJSON ApiSelectCoinsAction where
     parseJSON = genericParseJSON defaultRecordTypeOptions
-instance EncodeAddress n => ToJSON (ApiSelectCoinsData n) where
+instance ToJSON ApiSelectCoinsAction where
     toJSON = genericToJSON defaultRecordTypeOptions
 
+instance DecodeAddress n => FromJSON (ApiSelectCoinsPayments n) where
+    parseJSON = genericParseJSON defaultRecordTypeOptions
+instance EncodeAddress n => ToJSON (ApiSelectCoinsPayments n) where
+    toJSON = genericToJSON defaultRecordTypeOptions
+
+instance DecodeAddress n => FromJSON (ApiSelectCoinsData n) where
+    parseJSON = withObject "DelegationAction" $ \o -> do
+        p <- o .:? "payments"
+        a <- o .:? "delegation_action"
+        case (p, a) of
+            (Just _, Just _) -> fail "Specified both payments and action, pick one"
+            (Nothing, Just v) ->
+                pure $ ApiSelectForAction $ ApiSelectCoinsAction v
+            (Just v, Nothing) ->
+                pure $ ApiSelectForPayment $ ApiSelectCoinsPayments v
+            _ -> fail "No valid parse"
+instance EncodeAddress n => ToJSON (ApiSelectCoinsData n) where
+    toJSON (ApiSelectForPayment v) = toJSON v
+    toJSON (ApiSelectForAction v) = toJSON v
+
 instance DecodeAddress n => FromJSON (ApiCoinSelection n) where
     parseJSON = genericParseJSON defaultRecordTypeOptions
 instance EncodeAddress n => ToJSON (ApiCoinSelection n) where
     toJSON = genericToJSON defaultRecordTypeOptions
 
+apiCertificateOptions :: Aeson.Options
+apiCertificateOptions = Aeson.defaultOptions
+      { constructorTagModifier = camelTo2 '_'
+      , tagSingleConstructors = True
+      , fieldLabelModifier = camelTo2 '_' . dropWhile (== '_')
+      , omitNothingFields = True
+      , sumEncoding = TaggedObject
+          {
+            tagFieldName = "delegation_type"
+          , contentsFieldName = "contents"
+          }
+      }
+
+instance FromJSON ApiCertificate where
+    parseJSON = genericParseJSON apiCertificateOptions
+
+instance ToJSON ApiCertificate where
+    toJSON = genericToJSON apiCertificateOptions
+
+instance FromJSON (ApiT DelegationAction) where
+    parseJSON = withObject "DelegationAction" $ \o ->
+        o .: "action" >>= \case
+            "join" -> do
+                pid <- o .: "pool"
+                pure (ApiT $ Join (getApiT pid))
+            "quit" -> pure $ ApiT Quit
+            val -> fail ("Unexpeced action value: " <> T.unpack val)
+
+instance ToJSON (ApiT DelegationAction) where
+    toJSON (ApiT (RegisterKeyAndJoin _)) = error "RegisterKeyAndJoin not valid"
```

Comment:
That we won't have `RegisterKeyAndJoin` as a possible input in the API should not leak up here. May we not make this function partial and simply have a valid instance here :pray: 

The application logic should handle this properly. (or we should use a different type, more constrained that doesn't make it possible).

Reply:
> That we won't have RegisterKeyAndJoin as a possible input in the API should not leak up here.

Why not? This instance is for `ApiT DelegationAction`, not `DelegationAction`. So it's about our API.

Reply:
We want to avoid partial functions when we can. There's no reason for this function to be partial. That the application logic (even only in the API) do not accept `RegisterKeyAndJoin` as a constructor should not influence the JSON instance itself. We are actually _abusing_ this type in the first place here in the application logic. The proper way to address this would be to have an actual type that captures only the allowed actions; because it's convenient we use the internal type `DelegationAction` wrapped in an `ApiT` but let's not shoot ourselves in the foot on top by making the JSON instances partial